src/AvaloniaEdit.TextMate/TextEditorModel.cs

```diff
@@ -66,33 +69,19 @@ private void TextView_ScrollOffsetChanged(object sender, EventArgs e)
             }
         }
 
-        private void DocumentOnLineCountChanged(object sender, EventArgs e)
-        {
-            try
-            {
-                lock (_lock)
-                {
-                    _lineCount = _document.LineCount;
-                }
-            }
-            catch (Exception ex)
-            {
-                _exceptionHandler?.Invoke(ex);
-            }
-        }
-
         private void DocumentOnChanging(object sender, DocumentChangeEventArgs e)
         {
             try
             {
-                if (e.RemovedText is { })
+                if (e.RemovalLength > 0)
                 {
                     var startLine = _document.GetLineByOffset(e.Offset).LineNumber - 1;
                     var endLine = _document.GetLineByOffset(e.Offset + e.RemovalLength).LineNumber - 1;
 
                     for (int i = endLine; i > startLine; i--)
                     {
                         RemoveLine(i);
+                        _documentSnapshot.LineCount--;
```

Comment:
Why doesnt `_documentSnapshot.Update(e)` do this?

Reply:
This happens before the document is actually modified.

Reply:
Long answer:

We need to remove lines from the model in the `DocumentChanging` event (they cannot be processed in the `DocumentChanged` event because we would be accessing deleted offsets).

The `_documentSnapshot.Update(e)` is processed in. the `DocumentChanged`, so unfortunately we need to separate that logic.

So summarizing:
- We must process removed lines in the  `DocumentChanging` event.
- In the time-lapse between we remove lines to the model, and we actually update the line ranges, we need to actually say the `TMModel` we have less lines, until the `DocumentSnapshot` is fully updated.
- This is the reason we need to decrement the LineCount. Otherwise the `TMModel` would be accessing indexes out of range in the tokenizer thread.

Reply:
ok